# We run the pulsar services individually as OAuth2 doesn't seem to work in
# standalone mode. OAuth2 is also only available from pulsar v3 onwards. We use
# host networking as the mock OAuth server uses the SNI host name to determine
# the host name of the OAuth2 issuer URL, so we need to have the mock server
# reachable by localhost from both the pulsar services and the host network.
version: '3'
services:
  # Start zookeeper
  zookeeper:
    image: apachepulsar/pulsar:3.0.0
    container_name: zookeeper
    restart: on-failure
    network_mode: "host"
    environment:
      - metadataStoreUrl=zk:localhost:2181
      - metricsProvider.httpPort=7000
      - PULSAR_MEM=-Xms256m -Xmx256m -XX:MaxDirectMemorySize=256m
    command: >
      bash -c "bin/apply-config-from-env.py conf/zookeeper.conf && \
             bin/generate-zookeeper-config.sh conf/zookeeper.conf && \
             exec bin/pulsar zookeeper"
    healthcheck:
      test: ["CMD", "bin/pulsar-zookeeper-ruok.sh"]
      interval: 1s
      timeout: 5s
      retries: 300

  # Init cluster metadata
  pulsar-init:
    container_name: pulsar-init
    image: apachepulsar/pulsar:3.0.0
    network_mode: "host"
    env_file:
      - ./pulsar_auth-oauth2.conf
    command: >
      bin/pulsar initialize-cluster-metadata \
               --cluster cluster-a \
               --zookeeper localhost:2181 \
               --configuration-store localhost:2181 \
               --web-service-url http://localhost:8080 \
               --broker-service-url pulsar://localhost:6650
    depends_on:
      zookeeper:
        condition: service_healthy

  # Start bookie
  bookie:
    image: apachepulsar/pulsar:3.0.0
    container_name: bookie
    restart: on-failure
    network_mode: "host"
    environment:
      - clusterName=cluster-a
      - zkServers=localhost:2181
      - metadataServiceUri=metadata-store:zk:localhost:2181
      # otherwise every time we run docker compose uo or down we fail to start due to Cookie
      # See: https://github.com/apache/bookkeeper/blob/405e72acf42bb1104296447ea8840d805094c787/bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Cookie.java#L57-68
      - advertisedAddress=localhost
      - BOOKIE_MEM=-Xms512m -Xmx512m -XX:MaxDirectMemorySize=256m
    env_file:
      - ./pulsar_auth-oauth2.conf
    volumes:
      - "{{ .TmpDir }}:/pulsar/conf/dapr"
    depends_on:
      zookeeper:
        condition: service_healthy
      pulsar-init:
        condition: service_completed_successfully
    command: bash -c "bin/apply-config-from-env.py conf/bookkeeper.conf && exec bin/pulsar bookie"

  # Start broker
  broker:
    image: apachepulsar/pulsar:3.0.0
    container_name: broker
    restart: on-failure
    network_mode: "host"
    env_file:
      - ./pulsar_auth-oauth2.conf
    volumes:
      - "{{ .TmpDir }}:/pulsar/conf/dapr"
    environment:
      - metadataStoreUrl=zk:localhost:2181
      - zookeeperServers=localhost:2181
      - clusterName=cluster-a
      - managedLedgerDefaultEnsembleSize=1
      - managedLedgerDefaultWriteQuorum=1
      - managedLedgerDefaultAckQuorum=1
      - advertisedAddress=localhost
      - advertisedListeners=external:pulsar://127.0.0.1:6650
      - PULSAR_MEM=-Xms512m -Xmx512m -XX:MaxDirectMemorySize=256m
    depends_on:
      zookeeper:
        condition: service_healthy
      bookie:
        condition: service_started
    command: bash -c "bin/apply-config-from-env.py conf/broker.conf && exec bin/pulsar broker"
